Implement reverse geolocation via /findNearbyPlaceNameJSON endpoint

mfilej opened the pull request:
Adds `reverse_find` and `reverse_search` endpoints to client.

The refactoring should allow us to handle any errors returned by API in a sane manner.

mfilej committed:
Extract geonames API calls into module

mfilej committed:
Implement reverse geolocation

mfilej committed:
Make WebServices dumber

mfilej committed:
Dry up main endpoints a bit

mfilej committed:
Move param name translation to appropriate layer

mfilej committed:
No need to repeat max rows parameter value

mfilej committed:
Remove some more duplication

mfilej committed:
Add #reverse_search endpoint for symmetry

mfilej commented on lib/gemonames.rb:
```diff
@@ -36,17 +37,43 @@ def initialize(connection)
+    def reverse_find(latitude:, longitude:)
+      perform_find :find_nearby_place_name,
+        lat: latitude, lng: longitude
+    end
```
I guess I should comment on the above changes:

I tried to achieve a shape where it is easily visible which endpoints are called on the geonames API and which parameters are passed exactly (that's why I changed to the ugly `maxRows` for example).

mfilej committed:
Logging spec example is not related to search

mfilej committed:
Require spec_helper via .rspec file

mfilej committed:
Split ApiClient into own file

mfilej committed:
Reorder stuff

tomekw commented on lib/gemonames.rb:
```diff
@@ -36,17 +37,43 @@ def initialize(connection)
+
+    private
+
+    def perform_search(endpoint, **args)
```
Keyword arg for `endpoint`? `perform_search(:search, {})` reads weird :)

mfilej replied:
True. :smile: Maybe I can think of another name.

tomekw commented on lib/gemonames.rb:
```diff
@@ -36,17 +37,43 @@ def initialize(connection)
-      perform_search_request(
-        query: query, country_code: country_code, max_rows: limit
-      ).fetch("geonames").map { |result|
+      perform_search :search,
```
Parentheses around arguments (everywhere) :)

mfilej replied:
I'd prefer to skip them as they would add 2 additional lines to each method.

tomekw commented on lib/gemonames.rb:
```diff
@@ -36,17 +37,43 @@ def initialize(connection)
-        query: query, country_code: country_code, max_rows: 1
-      ).fetch("geonames")
+    def perform_find(endpoint, **args)
+      args = args.merge(maxRows: 1)
```
Maybe just inline it on line 75?

tomekw commented on lib/gemonames.rb:
```diff
@@ -36,17 +37,43 @@ def initialize(connection)
+        WebServices.public_send(endpoint, connection, **args)
+      )
+
+      results.map { |result|
```
One liner?

tomekw commented on lib/gemonames.rb:
```diff
@@ -36,17 +37,43 @@ def initialize(connection)
+    private
+
+    def perform_search(endpoint, **args)
+      results = extract_payload(
```
`results` variable not needed? Just chain it with `map`?

mfilej commented:
@tomekw re: inlining - I left the code in this shape so that it will be easier to refactor.

![final form](http://memecrunch.com/meme/310S2/this-isn-t-even-my-final-form/image.png)

mfilej committed:
Fix typo

mfilej committed:
Remove option to make cilent with given connection

mfilej committed:
Reimplement find using search

mfilej committed:
Rename #result? to #present?

But keep old one for backwards compatibility.

mfilej committed:
Remove rubocop method length restriction

It does not make sense for it to fail when a method only initializes a
big hash or an object with a lot of parameters.

mfilej committed:
Rename #perform_search to #perform

mfilej commented:
@tomekw I think the refactoring addressed all your comments (except parentheses).

mfilej committed:
Keep cassette naming consistent

RadekMolenda commented:
:star:

tomekw commented:
:star:
